File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Samoa.m
//
//  UIImage+Samoa.m
//  CountriesFlags
//

#import "Countries/UIImage+Samoa.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Samoa)

+ (UIImage *)flagSamoaWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:0.0f
                                         green:43.0f/255.0f
                                          blue:127.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorRed = [UIColor colorWithRed:206.0f/255.0f
                                        green:17.0f/255.0f
                                         blue:38.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width / 2.0f, size.height / 2.0f));
    
    [colorWhite setFill];
    // left
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 288.0f * 50.0f, size.height / 144.0f * 29.5f)
                        radius:size.height / 144.0f * 8.5f
                          type:FlagStarTypeFivePointed] fill];
    // top
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 288.0f * 72.0f, size.height / 144.0f * 11.5f)
                        radius:size.height / 144.0f * 8.5f
                          type:FlagStarTypeFivePointed] fill];
    // right
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 288.0f * 94.5f, size.height / 144.0f * 26.0f)
                        radius:size.height / 144.0f * 8.0f
                          type:FlagStarTypeFivePointed] fill];
    // bottom
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 288.0f * 72.0f, size.height / 144.0f * 60.5f)
                        radius:size.height / 144.0f * 10.5f
                          type:FlagStarTypeFivePointed] fill];
    // mid-right
    [[UIBezierPath starAtPoint:CGPointMake(size.width / 288.0f * 83.0f, size.height / 144.0f * 41.75f)
                        radius:size.height / 144.0f * 5.0f
                          type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Samoa.swift
//
//  UIImage+Samoa.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Samoa "wikipedia"
     
     Generates the flag of Samoa.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Samoa.
     */
    internal static func flagSamoa(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 0.0,
                                green: 43.0 / 255.0,
                                blue: 127.0 / 255.0,
                                alpha: 1.0)
        let colorWhite = UIColor.white
        let colorRed = UIColor(red: 206.0 / 255.0,
                               green: 17.0 / 255.0,
                               blue: 38.0 / 255.0,
                               alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size.halfSize))

        colorWhite.setFill()
        // left
        UIBezierPath.star(at: CGPoint(x: size.width / 288.0 * 50.0, y: size.height / 144.0 * 29.5),
                          radius: size.height / 144.0 * 17.0 / 2.0).fill()
        // top
        UIBezierPath.star(at: CGPoint(x: size.width / 288.0 * 72.0, y: size.height / 144.0 * 11.5),
                          radius: size.height / 144.0 * 17.0 / 2.0).fill()
        // right
        UIBezierPath.star(at: CGPoint(x: size.width / 288.0 * 94.5, y: size.height / 144.0 * 26.0),
                          radius: size.height / 144.0 * 16.0 / 2.0).fill()
        // bottom
        UIBezierPath.star(at: CGPoint(x: size.width / 288.0 * 72.0, y: size.height / 144.0 * 60.5),
                          radius: size.height / 144.0 * 21.0 / 2.0).fill()
        // mid-right
        UIBezierPath.star(at: CGPoint(x: size.width / 288.0 * 83.0, y: size.height / 144.0 * 41.75),
                          radius: size.height / 144.0 * 10.0 / 2.0).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
